### ISSUE TITLE
[proposal] ActiveSearch - Rails Search 

### ISSUE THREAD

**Issue by yagudaev** [rails/rails#40989]
# Overview

Over the last two years, I've had the pleasure of doing and re-doing our search at work 4 times. Search and information retrieval is a deep and complex field. Nevertheless, I feel like it should be easy to add a simple search to a rails website.

I recently discovered that [Django already solved this problem](https://docs.djangoproject.com/en/3.1/topics/db/search/). Thus, Rails can do it too and even improve upon it.

# Proposed Approach

Similar to a lot of Rails source, we already have something working in production @ [Ollie Order](https://www.ollieorder.com/). However, it can be a lot better if it was deeply integrated into rails. The recent changes that @eileencodes made to support multiple database can be key to paving support for more advanced search solution like ElasticSearch, Algolia, etc. Baking a lot of things right into ActiveRecord would be ideal.

For the first version, I'd like to focus on keeping it simple using a single database for data & searching. Most databases these days support some form of full-text search. Similar to ActiveRecord it will support the ability to use any search engine you want. Unlike ActiveRecord, it will allow users to choose the search engine per view. Different views of the same data need different search capabilities and performance (e.g. /admin/posts vs /posts will get different traffic).

Let's start with a visual of a modern search:

![Screen Shot 2020-12-31 at 8 33 26 AM](https://user-images.githubusercontent.com/1386966/103418404-55d75700-4b43-11eb-9afd-7084e4f9a5aa.png)
(source: [Elastic App Search UI](https://www.elastic.co/enterprise-search/search-ui?q=mountain&size=n_6_n))

Everything above can be accomplished by stringing together several gems. There is a natural coupling between all these components, making it a pain to integrate individually. Let's break it down:

1. Filtering - doable with AR, but a gem like [Ransack](https://github.com/activerecord-hackery/ransack) makes building the UI and persisting searches in the URL far easier
2. Free-text search - no AR implementation, either raw SQL or use a gem like [pg_search](https://github.com/Casecommons/pg_search)
3. Ranking -  no AR implementation, either raw SQL or use a gem like [pg_search](https://github.com/Casecommons/pg_search)
4. Sorting - doable with AR, but to sort by relevance see point 3
5. Facets - write custom AR code to do group_bys or use a gem like [forty_facets](https://github.com/FortySource/forty_facets)
6. Pagination - no AR implementation, use [Pagy](https://github.com/ddnexus/pagy) or a similar gem
7. Highlighting - no AR implementation, use [pg_search](https://github.com/Casecommons/pg_search) or implement your own string lookups + annotation
8. Multi-model search - search across multiple objects types, e.g. global search. No AR implementation, use [pg_search](https://github.com/Casecommons/pg_search).
9. Authentication/Authorization - certain data can only be viewed if you are authenticated and authorized. A combination of devise and [pundit](https://github.com/varvet/pundit) is the usual suspects here

# Proposed API

Mostly TBD (will update as the discussion here evolves).

# Ask

Looking to hear how others approach & solve this problem in Rails. I'm also looking for a few individuals who are equally passionate and want to help solve this problem.

# Resources

## Similar APIs
* [Django Search API](https://docs.djangoproject.com/en/3.1/topics/db/search/)
* [Laravel Search API](https://laravel.com/docs/8.x/scout)

## Full-text search
* [Rails Cast Full Text Search PG](http://railscasts.com/episodes/343-full-text-search-in-postgresql)
* [Algolia Stack](https://stackshare.io/posts/how-algolia-built-their-realtime-search-as-a-service-product)
* [MySQL Full Text Search](https://dev.mysql.com/doc/refman/8.0/en/fulltext-search.html)
* [Postgres Full Text Search](https://www.postgresql.org/docs/current/textsearch.html)
* [SQLite Full Text Search](https://sqlite.org/fts5.html)
* [Redis Full Text Search](https://redislabs.com/modules/redis-search/)
* [MS SQL Full Text Search](https://docs.microsoft.com/en-us/sql/relational-databases/search/query-with-full-text-search?view=sql-server-ver15)

# Next Steps

* I'll be putting together all these gems in a repo and a rails guide document to experiment with what this new API might look like.

This issue will be updated with progress. So check in occasionally.

**Comment by eileencodes** [comment]
Hey @yagudaev I appreciate the thorough issue writeup and enthusiasm but we don't take feature requests or proposals on the issues tracker. We reserve it for bugs in the framework. The [message board](https://discuss.rubyonrails.org/) is the right place to propose this and discuss an API. Ideally the best path forward would be to build a gem that extends Active Record and then ask if we (the Core team) wants this feature in Rails. That way we would be able to interact with the API. I'm unsure whether we would want to include this in AR, but we can discuss that over in the forum. Thanks for understanding!

**Comment by yagudaev** [comment]
@eileencodes thanks for the explanation :).

I've created a discussion here: https://discuss.rubyonrails.org/t/feature-proposal-activesearch-rails-search/76834 on the subject of search.

I've had to take a step back to understand why I choose to post this as a github issue and why it might not make sense as you said. I've created a separate discussion here around that important topic: https://discuss.rubyonrails.org/t/community-engagement-discussion-feature-requests-roadmap/76836
